vm/systemSmartContracts/validator.go

```diff
@@ -1069,46 +1075,73 @@ func (v *validatorSC) stake(args *vmcommon.ContractCallInput) vmcommon.ReturnCod
 		}
 	}
 
-	if !v.isNumberOfNodesTooHigh(registrationData) {
-		v.activateStakingFor(
-			blsKeys,
-			registrationData,
-			validatorConfig.NodePrice,
-			registrationData.RewardAddress,
-			args.CallerAddr,
-		)
-	} else {
-		numRegisteredBlsKeys := int64(len(registrationData.BlsPubKeys))
+	v.activateNewBLSKeys(registrationData, blsKeys, newKeys, &validatorConfig, args)
+
+	err = v.saveRegistrationData(args.CallerAddr, registrationData)
+	if err != nil {
+		v.eei.AddReturnMessage("cannot save registration data: error " + err.Error())
+		return vmcommon.UserError
+	}
+
+	return vmcommon.Ok
+}
+
+func (v *validatorSC) activateNewBLSKeys(
+	registrationData *ValidatorDataV2,
+	blsKeys [][]byte,
+	newKeys [][]byte,
+	validatorConfig *ValidatorConfig,
+	args *vmcommon.ContractCallInput,
+) {
+	numRegisteredBlsKeys := len(registrationData.BlsPubKeys)
+	numNodesTooHigh := v.activateStakingFor(
```

Comment:
It is a bit confusing having a function called `activateStakingFor` -> which **should do** something, but returns true when it **doesn't do** it (activate), e.g.: number of nodes is too high). Perhaps we can either change return logic of variable/func names

Reply:
pushed